fem/fe_coll.hpp

```diff
@@ -66,6 +66,17 @@ class FiniteElementCollection
 
    virtual int GetContType() const = 0;
 
+   /** @note The following methods provide the same information as the
+       corresponding methods of the FiniteElement base class.
+       @{
+   */
+   virtual int GetRangeType() const = 0;
+   virtual int GetDerivRangeType() const = 0;
+   virtual int GetMapType() const = 0;
+   virtual int GetDerivType() const = 0;
+   virtual int GetDerivMapType() const = 0;
+   /** @} */
+
```

Comment:
These methods seem redundant. Why not simply query these using the `FiniteElement` returned by `FiniteElementSpace::GetFE(0)`?

Reply:
I thought of that but we occasionally have partitions with no elements. This may be particularly common on a `SubMesh` because the partitions are determined by the full mesh.

Another way to avoid this redundancy would be to store pointers to the `FiniteElement` objects in the `FiniteElementCollection` base class. Most, if not all, derived classes do store `FiniteElement` objects for various geometry types but they don't store them in a uniform fashion. I briefly considered making this more uniform but the changes would be very extensive.

Reply:
Implementing these methods in the base class required a dimension argument because this is not stored in the base class. I considered storing `dim` in the base class but some collections are designed for arbitrary dimension so it's better not to alter this.

In the unlikely event that a suitable `FiniteElement` cannot be found I wanted to be able to return a meaningful value from these methods. Just returning some arbitrarily chosen default could lead to trouble if someone depends on some particular behavior. Therefore I added two enumeration values `UNKNOWN_RANGE_TYPE` and `UNKNOWN_MAP_TYPE`. For `DerivType` the existing value of `NONE` is probably sufficient to avoid unexpected behavior. I've set these new identifiers to have values of `-1` so that existing code which uses the integer values will not need to be altered.

Reply:
Looks good, thanks @mlstowell!

Reply:
Thank you. That was an excellent suggestion, @v-dobrev